packages/backend-common/src/tokens/ServerTokenManager.ts

```diff
@@ -86,52 +112,88 @@ export class ServerTokenManager implements TokenManager {
       );
     }
+
     if (this.privateKeyPromise) {
       return this.privateKeyPromise;
     }
+
     const promise = (async () => {
-      const secret = await generateSecret('HS256');
+      const secret = await generateSecret(TOKEN_ALG);
       const jwk = await exportJWK(secret);
       this.verificationKeys.push(base64url.decode(jwk.k ?? ''));
       this.signingKey = this.verificationKeys[0];
       return;
     })();
 
     try {
-      // If we fail to generate a new key, we need to clear the state so that
-      // the next caller will try to generate another key.
+      this.privateKeyPromise = promise;
       await promise;
     } catch (error) {
-      this.logger.error(`Failed to generate new key, ${error}`);
+      // If we fail to generate a new key, we need to clear the state so that
+      // the next caller will try to generate another key.
+      this.options.logger.error(`Failed to generate new key, ${error}`);
       delete this.privateKeyPromise;
     }
+
     return promise;
   }
 
   async getToken(): Promise<{ token: string }> {
     if (!this.verificationKeys.length) {
       await this.generateKeys();
     }
-    const sub = 'backstage-server';
-    const jwt = await new SignJWT({ alg: 'HS256' })
-      .setProtectedHeader({ alg: 'HS256', sub: sub })
-      .setSubject('backstage-server')
-      .sign(this.signingKey);
-    return { token: jwt };
+
+    if (this.currentTokenPromise) {
+      return this.currentTokenPromise;
+    }
+
+    const result = Promise.resolve().then(async () => {
+      const jwt = await new SignJWT({})
+        .setProtectedHeader({ alg: TOKEN_ALG })
+        .setSubject(TOKEN_SUB)
+        .setExpirationTime(DateTime.now().plus(TOKEN_EXPIRY).toUnixInteger())
+        .sign(this.signingKey);
+      return { token: jwt };
+    });
+
+    this.currentTokenPromise = result;
+
+    result
+      .then(() => {
+        setTimeout(() => {
+          this.currentTokenPromise = undefined;
+        }, TOKEN_EXPIRY.minus({ minutes: 5 }).toMillis());
+      })
+      .catch(() => {
+        this.currentTokenPromise = undefined;
+      });
+
+    return result;
   }
 
   async authenticate(token: string): Promise<void> {
     let verifyError = undefined;
+
     for (const key of this.verificationKeys) {
       try {
-        await jwtVerify(token, key);
-        // If the verify succeeded, return
+        const result = await jwtVerify(token, key);
+        if (result.protectedHeader.alg !== TOKEN_ALG) {
+          throw new NotAllowedError(
+            `Illegal alg "${result.protectedHeader.alg}"`,
+          );
+        }
+        if (result.payload.sub !== TOKEN_SUB) {
+          throw new NotAllowedError(`Illegal sub "${result.payload.sub}"`);
+        }
+        // TODO(freben): Reject missing payload.exp in the future as well
+        // The jose library does NOT throw if exp is not set in the token, but DOES throw if exp is set and expired
```

Comment:
Think we could get this through somehow as well? Otherwise we'll end up with yet another breaking change in the future.

Or perhaps we keep this all as is, call this a deprecation, and not a breaking change? The only breakage right now is for consumers that were calling `getToken` and holding on to it. If we go down that route, perhaps we can log a deprecation warning rather than throw an error if the token is expired? And then make that an error and require expiration in the following release. Should be able to get around `jose` by setting clock tolerance to `Infinity`